base/resolver_node_lookup.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package base
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+const lookupTimeout = time.Second * 3
+
+// nodeLookupResolver implements Resolver.
+// It queries http(s)://<address>/_status/nodes and extracts the first node's address.
+// This is useful for http load balancers which will not foward RPC.
+// It is never exhausted.
+type nodeLookupResolver struct {
+	context   Context
+	typ       string
+	addr      string
+	exhausted bool
+	// We need our own client so that we may specify timeouts.
+	httpClient *http.Client
+}
+
+// Type returns the resolver type.
+func (nl *nodeLookupResolver) Type() string { return nl.typ }
+
+// Addr returns the resolver address.
+func (nl *nodeLookupResolver) Addr() string { return nl.addr }
+
+// GetAddress returns a net.Addr or error.
+// Upon errors, we set exhausted=true, then flip it back when called again.
```

Comment:
It's because of the weird startup issue in single-node mode. The http server doesn't start until we've either resolved a gossip bootstrap address, or exhausted them all. I get around that by setting the resolver to exhausted on errors, and returning right away if we are exhausted. The caller runs GetAddress() before checking IsExhausted().

I still want to rework this part, it's getting really messy.